rhp/v4/rpc.go

```diff
@@ -0,0 +1,529 @@
+	}
+
+	// TODO: stream?
+	var sector [rhp4.SectorSize]byte
+	copy(sector[:], data)
+	root := rhp4.SectorRoot(&sector)
+	if root != resp.Root {
+		return types.Hash256{}, fmt.Errorf("invalid root returned: expected %v, got %v", root, resp.Root)
+	}
+	return root, nil
+}
+
+// RPCModifySectors modifies sectors on the host
+func RPCModifySectors(t TransportClient, cs consensus.State, prices rhp4.HostPrices, sk types.PrivateKey, contract ContractRevision, actions []rhp4.WriteAction) (types.V2FileContract, error) {
+	s := t.DialStream()
+	defer s.Close()
+
+	req := rhp4.RPCModifySectorsRequest{
+		ContractID: contract.ID,
+		Prices:     prices,
+		Actions:    actions,
+	}
+	req.ChallengeSignature = sk.SignHash(req.ChallengeSigHash(contract.Revision.RevisionNumber + 1))
+
+	if err := rhp4.WriteRequest(s, rhp4.RPCModifySectorsID, &req); err != nil {
+		return types.V2FileContract{}, fmt.Errorf("failed to write request: %w", err)
+	}
+
+	var resp rhp4.RPCModifySectorsResponse
+	if err := rhp4.ReadResponse(s, &resp); err != nil {
+		return types.V2FileContract{}, fmt.Errorf("failed to read response: %w", err)
+	}
+
+	revision, err := rhp4.ReviseForModifySectors(contract.Revision, req, resp)
+	if err != nil {
+		return types.V2FileContract{}, fmt.Errorf("failed to revise contract: %w", err)
+	}
+
+	sigHash := cs.ContractSigHash(revision)
+	revision.RenterSignature = sk.SignHash(sigHash)
+
+	signatureResp := rhp4.RPCModifySectorsSecondResponse{
+		RenterSignature: revision.RenterSignature,
+	}
+	if err := rhp4.WriteResponse(s, &signatureResp); err != nil {
+		return types.V2FileContract{}, fmt.Errorf("failed to write signature response: %w", err)
+	}
+
+	var hostSignature rhp4.RPCModifySectorsThirdResponse
+	if err := rhp4.ReadResponse(s, &hostSignature); err != nil {
+		return types.V2FileContract{}, fmt.Errorf("failed to read host signatures: %w", err)
+	}
+	// validate the host signature
+	if !contract.Revision.HostPublicKey.VerifyHash(sigHash, hostSignature.HostSignature) {
+		return contract.Revision, rhp4.ErrInvalidSignature
+	}
+	// return the signed revision
+	return revision, nil
+}
+
+// RPCFundAccounts funds accounts on the host
+func RPCFundAccounts(t TransportClient, cs consensus.State, sk types.PrivateKey, contract ContractRevision, deposits []rhp4.AccountDeposit) (types.V2FileContract, []types.Currency, error) {
+	var total types.Currency
+	for _, deposit := range deposits {
+		total = total.Add(deposit.Amount)
+	}
+	revision, err := rhp4.ReviseForFundAccount(contract.Revision, total)
+	if err != nil {
+		return types.V2FileContract{}, nil, fmt.Errorf("failed to revise contract: %w", err)
+	}
+	sigHash := cs.ContractSigHash(revision)
+	revision.RenterSignature = sk.SignHash(sigHash)
+
+	req := rhp4.RPCFundAccountsRequest{
+		ContractID:      contract.ID,
+		Deposits:        deposits,
+		RenterSignature: revision.RenterSignature,
+	}
+
+	s := t.DialStream()
+	defer s.Close()
```

Comment:
I would prefer to add context to it for cancellation. We may be able to get away with adding it to the transport. This would probably be fine for the mux with no changes:
```go
func (mt *muxTransport) DialStream(ctx context.Context) net.Conn {
  s := mt.mux.DialStream()
  go func() {
	<-ctx.Done()
	s.Close()
  }()
  return s
}
```

Would need to check that closing the stream does actually interupt any calls to read and write.

Reply:
Mux v2 has `DialStreamContext`